sql/src/main/java/io/crate/operation/aggregation/impl/PercentileAggregation.java

```diff
@@ -0,0 +1,263 @@
+        }
+
+        @Override
+        public int compareValueTo(TDigestState val1, TDigestState val2) {
+            return 0;
+        }
+    }
+
+    public static class TDigestState extends AVLTreeDigest {
+
+        private final double compression;
+        private final double[] fractions;
+
+        public TDigestState(double compression, double[] fractions) {
+            super(compression);
+            this.compression = compression;
+            this.fractions = fractions;
+        }
+
+        @Override
+        public double compression() {
+            return compression;
+        }
+
+        public double[] fractions() {
+            return fractions;
+        }
+
+        public static void write(TDigestState state, StreamOutput out) throws IOException {
+            out.writeDouble(state.compression);
+            out.writeDoubleArray(state.fractions);
+            out.writeVInt(state.centroidCount());
+            for (Centroid centroid : state.centroids()) {
+                out.writeDouble(centroid.mean());
+                out.writeVLong(centroid.count());
+            }
+        }
+
+        public static TDigestState read(StreamInput in) throws IOException {
+            double compression = in.readDouble();
+            double[] fractions = in.readDoubleArray();
+            TDigestState state = new TDigestState(compression, fractions);
+            int n = in.readVInt();
+            for (int i = 0; i < n; i++) {
+                state.add(in.readDouble(), in.readVInt());
+            }
+            return state;
+        }
+
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    @Nullable
+    @Override
+    public TDigestState newState(RamAccountingContext ramAccountingContext) {
+        return null;
+    }
+
+    @Override
+    public TDigestState iterate(RamAccountingContext ramAccountingContext, TDigestState state, Input... args) throws CircuitBreakingException {
+        if (state == null) {
+            double[] fractions;
+            if (args[1].value() != null) {
+                if (args[1].value().getClass().isArray()) {
+                    Object[] values = (Object[]) args[1].value();
+                    if (values.length == 0) {
+                        throw new IllegalArgumentException("fraction should be in range [0,1]");
+                    }
+                    fractions = toDoubleArray(values);
+                } else {
+                    fractions = new double[1];
+                    fractions[0] = DataTypes.DOUBLE.value(args[1].value());
+                }
+                state = new TDigestState(COMPRESSION, fractions);
+            } else {
+                state = new TDigestState(COMPRESSION, null);
+            }
+        }
+        Double value = DataTypes.DOUBLE.value(args[0].value());
+        if (value != null) {
+            state.add(value);
+        }
+        return state;
+    }
+
+    private double[] toDoubleArray(Object[] array) {
```

Comment:
Might also make sense to move this into DoubleType or ArrayType and make it public - to make it re-usable